### PR TITLE
Amd v9 p1

### PR DESCRIPTION
**Pull request by babumoger**
## Description
AMD is working on to support RDT feature(AMD calls it QoS feature) in next generation hardware. Right now the tool intel-cmt-cat does not support AMD. Majority of these features are similar in both the vendors. However, there are some differences in the way some of these features are implemented in each vendor. We needed to separate those functions and define the vendor specific versions of these functions. This pull request is a first attempt in that direction.

The specification for this AMD feature is available at
https://developer.amd.com/wp-content/resources/56375.pdf

## Affected parts

- [X] library
- [X] pqos utility
- [X] rdtset utility
- [ ] other: (please specify)

## Motivation and Context

Idea here is to have a one tool to support both Intel and AMD. It becomes easy to maintain one tool for both the vendors. It also benefits the OS vendors.

## How Has This Been Tested?

 This pull request only has partial set of changes. These patches does not add AMD support yet. So, I have tested on Intel box to make sure it does not introduce any regressions.
Hardware configuration of the Intel system.
# lscpu
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
Address sizes:       46 bits physical, 48 bits virtual
CPU(s):              32
On-line CPU(s) list: 0-31
Thread(s) per core:  2
Core(s) per socket:  16
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Gold 5218 CPU @ 2.30GHz
Stepping:            7
CPU MHz:             1001.145
CPU max MHz:         3900.0000
CPU min MHz:         1000.0000
BogoMIPS:            4600.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            22528K
NUMA node0 CPU(s):   0-31
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge
mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall
nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl
xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl
vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2
x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm
abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin
ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept
vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm
cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt
avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc
cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku
ospke avx512_vnni md_clear flush_l1d arch_capabilities

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
  User interface for the tool has not changed. So old documentation still holds good. But we may need to update the doc once the AMD support is fully integrated.

### CODE REVIEW

File: rdtset/rdt.c [hunk 1]
```diff
@@ -160,7 +160,7 @@ rdt_cfg_is_valid(const struct rdt_cfg cfg)
 
 	case PQOS_CAP_TYPE_MBA:
 		return cfg.u.mba != NULL && cfg.u.mba->mb_max > 0 &&
-			((cfg.u.mba->ctrl == 0 && cfg.u.mba->mb_max <= 100) ||
+			((cfg.u.mba->ctrl == 0 && cfg.u.mba->mb_max <= RDT_MAX_MBA) ||
```
Comment (aleksinx) [inline rdtset/rdt.c]: line over 80 characters

File: lib/pqos.h [hunk 2]
```diff
@@ -922,6 +922,19 @@ int pqos_mba_get(const unsigned socket,
...
+ *
+ * @param [in] cpu CPU information structure from \a pqos_cap_get
+ * @param [out] count place to store actual number of l3cat ids returned
+ *
+ * @return Allocated array of size \a count populated with l3cat id's
+ * @retval NULL on error
+ */
+unsigned *
+pqos_cpu_get_l3cat_ids(const struct pqos_cpuinfo *cpu,
+		    unsigned *count);
```
Comment (aleksinx) [inline lib/pqos.h]: Declarations should be aligned

File: lib/pqos.h [hunk 3]
```diff
@@ -922,6 +922,19 @@ int pqos_mba_get(const unsigned socket,
...
+ *
+ * @param [in] cpu CPU information structure from \a pqos_cap_get
+ * @param [out] count place to store actual number of mba ids returned
+ *
+ * @return Allocated array of size \a count populated with mba id's
+ * @retval NULL on error
+ */
+unsigned *
+pqos_cpu_get_mba_ids(const struct pqos_cpuinfo *cpu,
+		  unsigned *count);
```
Comment (aleksinx) [inline lib/pqos.h]: Declarations should be aligned

File: lib/allocation.c [hunk 4]
```diff
@@ -1431,21 +1431,21 @@ hw_alloc_reset(const enum pqos_cdp_config l3_cdp_cfg,
...
-                 * Change L3 COS definition on all sockets
+		 * Change L3 COS definition on all l3cat ids
                  * so that each COS allows for access to all cache ways
                  */
-                for (j = 0; j < sockets_num; j++) {
+		for (j = 0; j < l3cat_id_num; j++) {
                         unsigned core = 0;
 
-                        ret = pqos_cpu_get_one_core(m_cpu, sockets[j], &core);
+			ret = pqos_cpu_get_one_core(m_cpu, l3cat_ids[j], &core);
```
Comment (aleksinx) [inline lib/allocation.c]: pqos_cpu_get_one_core expect socket id as parameter.

File: lib/allocation.c [hunk 5]
```diff
@@ -1486,13 +1486,13 @@ hw_alloc_reset(const enum pqos_cdp_config l3_cdp_cfg,
...
-                 * Go through all sockets and reset MBA class defintions
+		 * Go through all L3 CAT ids and reset MBA class definitions
                  * 0 is the default MBA COS value in linear mode.
                  */
-                for (j = 0; j < sockets_num; j++) {
+		for (j = 0; j < l3cat_id_num; j++) {
                         unsigned core = 0;
 
-                        ret = pqos_cpu_get_one_core(m_cpu, sockets[j], &core);
+			ret = pqos_cpu_get_one_core(m_cpu, l3cat_ids[j], &core);
```
Comment (aleksinx) [inline lib/allocation.c]: pqos_cpu_get_one_core expect socket id as parameter.

File: lib/os_allocation.c [hunk 6]
```diff
@@ -1057,7 +1057,7 @@ verify_mba_id(const unsigned mba_id, const struct pqos_cpuinfo *cpu)
 }
 
 int
-os_l3ca_set(const unsigned socket,
+os_l3ca_set(const unsigned l3cat_id,
```
Comment (aleksinx) [inline lib/os_allocation.c]: Please update documentation of the function

File: rdtset/rdt.c [hunk 7]
```diff
@@ -251,10 +251,10 @@ get_max_res_id(unsigned technology)
                 max = num_ids > max ? num_ids : max;
                 free(ids);
         }
-        /* get number of sockets */
+	/* get number of l3cat_ids */
         if (technology & (1 << PQOS_CAP_TYPE_L3CA) ||
```
Comment (aleksinx) [inline rdtset/rdt.c]: In subsequent patch it should be updated to check max mba_id

File: lib/os_cap.c [hunk 8]
```diff
@@ -784,11 +784,11 @@ os_cap_get_mba_ctrl(const struct pqos_cap *cap,
                         if (ret != PQOS_RETVAL_OK)
                                 continue;
 
-                        for (i = 0; i < sock_num; i++) {
+			for (i = 0; i < mba_id_num; i++) {
                                 struct pqos_mba mba;
 
-                                ret = resctrl_schemata_mba_get(schmt, sock[i],
-                                                               &mba);
+				ret = resctrl_schemata_mba_get(schmt, mba_ids[i],
```
Comment (aleksinx) [inline lib/os_cap.c]: line over 80 characters

File: rdtset/rdt.c (hunk 1 above)
Comment (aleksinx) [reply rdtset/rdt.c]: This issue was reported by Travis. I have added more comments

File: lib/allocation.c (hunk 4 above)
Comment (aleksinx) [reply lib/allocation.c]: I would prefer to have this changes in the current pull request.

**Comment by babumoger** [comment]
Opened the new pull request AMD v9 pull 2 to handle these changes.

**Pull request closed without merging**